File: core/agodrain.py
#!/usr/bin/env python
#
# drain - output all messages of AMQP based automation control
#
# Copyright (c) 2012 Christoph Jaeger office@diakonesis.at>
#

import sys
import syslog
import ConfigParser, os

from qpid.messaging import *
from qpid.util import URL
from qpid.log import enable, DEBUG, WARN


config = ConfigParser.ConfigParser()
config.read('/etc/opt/agocontrol/config.ini')

try:
	username = config.get("system","username")
except:
	username = "agocontrol"

try:
	password = config.get("system","password")
except:
	password = "letmein"

try:
	broker = config.get("system","broker")
except:
	broker = "localhost"

try:
	debug = config.get("system","debug")
except:
	debug = "WARN"

if debug=="DEBUG":
	enable("qpid", DEBUG)
else:
	enable("qpid", WARN)

# route stderr to syslog
class LogErr:
	def write(self, data):
		syslog.syslog(syslog.LOG_ERR, data)

syslog.openlog(sys.argv[0], syslog.LOG_PID, syslog.LOG_DAEMON)
sys.stderr = LogErr()

connection = Connection(broker, username=username, password=password, reconnect=True)
try:
	connection.open()
	session = connection.session()
	receiver = session.receiver("agocontrol; {create: always, node: {type: topic}}")
	receiver.capacity=100
	while True:
		try:
			message = receiver.fetch()
			print message

			session.acknowledge()
		except Empty:
			pass
except SendError, e:
	print e
except ReceiverError, e:
	print e
except KeyboardInterrupt:
	pass
finally:
	connection.close()

File: core/drain/agodrain.cpp
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
using namespace std;

#include <qpid/messaging/Connection.h>
#include <qpid/messaging/Message.h>
#include <qpid/messaging/Receiver.h>
#include <qpid/messaging/Sender.h>
#include <qpid/messaging/Session.h>
#include <qpid/messaging/Address.h>
using namespace qpid::types;
using namespace qpid::messaging;

#include "agoclient.h"
using namespace agocontrol;

// qpid session and sender/receiver
Receiver receiver;
Sender sender;
Session session;
Connection *connection;


int main(int argc, char **argv) {
        string broker;

        Variant::Map connectionOptions;
        broker=getConfigOption("system", "broker", "localhost:5672");
        connectionOptions["username"]=getConfigOption("system", "username", "agocontrol");
        connectionOptions["password"]=getConfigOption("system", "password", "letmein");

        connectionOptions["reconnect"] = "true";

        connection = new Connection(broker, connectionOptions);
        try {
                connection->open(); 
                session = connection->createSession(); 
                receiver = session.createReceiver("agocontrol; {create: always, node: {type: topic}}"); 
                sender = session.createSender("agocontrol; {create: always, node: {type: topic}}"); 
        } catch(const std::exception& error) {
                std::cerr << error.what() << std::endl;
                connection->close();
                printf("could not startup\n");
                return 1;
        }


        while (true) {
                try{
                        Message message = receiver.fetch(Duration::SECOND * 3);
                        std::cout << "Message(properties=" << message.getProperties() << ", content='" ;
                        if (message.getContentType() == "amqp/map") {
                                Variant::Map map;
                                decode(message, map);
                                std::cout << map;
                        } else {
                                std::cout << message.getContent();
                        }
                        std::cout  << "')" << std::endl;
                        session.acknowledge(message);

                } catch(const NoMessageAvailable& error) {
                        
                } catch(const std::exception& error) {
                        std::cerr << error.what() << std::endl;
                        usleep(50);
                }
        }

}
